Fix if a source does not have a name set

isantop opened the pull request:
I believe the issue described in #49 is caused by some sources being added which don't have an explicit name set. When version 2.0.0 of Repoman checks over these sources, there is not an `X-Repolib-Name` key set, causing the `KeyError` described. This fixes this by wrapping that key in a `try: except:` statement and setting a fallback value.

Fixes #49 
Fixes pop-os/repoman#103

Commit:
fix(source): fix for getting sources with missing names

Commit:
chore(release): release 2.0.1

jacobgkau commented:
Confirmed the issue being described using a manually-created `.sources` file with no `X-Repolib-Name` line (which I assume could have also been created by older repolib versions before this key was mandatory?)

After the fix, I can launch Repoman again, and it treats the source with a missing name as if it has the auto-generated name. The file is not actually modified to include this name unless I open and save the source.

A brief question about the code, what is the point of the `if name:` block? If there was somehow not a name there, wouldn't the code fail anyway on the return statement? (It almost seems like that `if` block should be part of the `except` block, or that `if` statement should also be checking that `self['X-Repolib-Name']` doesn't exist like it used to; as it is now, we're setting `name` to `self['X-Repolib-Name']` and then setting `self['X-Repolib-Name']` right back to `name` redundantly.)

While testing this, I was able to create another crash case. After removing both `X-Repolib-Name` and `X-Repolib-ID` from a `.sources` file:

![Screenshot from 2022-10-06 09-51-59](https://user-images.githubusercontent.com/7199422/194360754-ef29acc1-eb99-4764-99f4-0b72e1d2cc2d.png)


And then launching Repoman and opening/saving that source, those keys are placed on the bottom:

![Screenshot from 2022-10-06 09-52-14](https://user-images.githubusercontent.com/7199422/194360769-a97a7350-0abe-4aca-a1ee-bf2bff906bc7.png)

If I place another line break before `X-Repolib-ID`:
![Screenshot from 2022-10-06 09-52-32](https://user-images.githubusercontent.com/7199422/194360784-08c33613-43fc-413a-bf9a-f8e8ac5e9491.png)

Then open repoman again, repoman crashes with `IndexError: list index out of range`, and I get a duplicate `X-Repolib-ID` in the file (this is without having opened/saved the source, since the crash happens before I could have done that):

![Screenshot from 2022-10-06 09-52-59](https://user-images.githubusercontent.com/7199422/194360797-8cdf1c8f-9a7f-4736-b51f-efb6decb9d49.png)

Also, I noticed that if I just set the `Name` field on a `.list` source to empty and save, repoman's repair attempt next time I launch it doesn't seem to work (and I keep getting an error dialogue on every launch):

![Screenshot from 2022-10-06 09-48-33](https://user-images.githubusercontent.com/7199422/194360564-22195f4b-debd-40ae-993c-cadcb1380d8d.png)

The issue presumably being that the name and ID comments are on a single line and/or that the name comment is empty.

Commit:
fix(source): fix for missing fallback condition

Commit:
fix(parsedeb): add fallback for blank names

Fixes a repoman crash when loading a legacy file with a blank name comment.

Commit:
fix(file): add a check for malformed sources and halt

Causes repoman to throw an error if a source file is malformed and
prevents repolib from processing the file, to avoid unexpected
behavior.

isantop commented:
It looks like I was missing a fallback condition which should have been present. 

For the issue with .list files, the problem was occurring because adding the `X-Repolib-Name` comment told repolib to expect a name, but then said name was blank. This shouldn't be valid, but it should be easy enough to add a simple fallback where if at the end of parsing the name is still blank, simply set it to the ident. This won't fix the issue if both the Name and Ident are blank, but in that case such a file is genuinely malformed and we shouldn't proceed without manual correction (repolib will should never save such a file, as the ident can't generally be changed once the source is created. That error message in Repoman indicates that repolib is ignoring that file.

For the first crash test case, that's happening because putting a new line in the middle of the source is an invalid DEB822 configuration. Sources in DEB822 are terminated by a new line, so you're effectively adding a new source which then doesn't have URIs or other components required to be a valid source. Repolib should not be saving files with that type of glitch in them, but it should also fail on parsing them (and trigger the error message from the other crash informing the user of this issue). I did notice that the command to run to get information is wrong, but that will need to be fixed in Repoman. I did add a check to source-file loading triggering that when a file has errors like that, which will also cause Repolib to ignore the file when processing it.

jacobgkau reviewed (approved):
I now correctly get a `Some sources contained errors` dialogue with the bad extra line break instead of a crash. The fix for the no-name crash still works.

For the legacy `.list` format, I no longer get a blank name saved to the file. I am still seeing that if I have a file start with a comment above the `deb` lines:

![Screenshot from 2022-10-06 15-27-13](https://user-images.githubusercontent.com/7199422/194422389-80acd2c6-76cb-49b3-bbed-cc16e8e2c082.png)

...and I simply open and save that source, then the comments are moved to the `deb` lines (and duplicated on each line):

![Screenshot from 2022-10-06 15-27-30](https://user-images.githubusercontent.com/7199422/194422391-d791a8df-0d76-416d-a253-8a1185eef8e6.png)

I think this is expected as of https://github.com/pop-os/repolib/pull/47, but just wanted to call it out to be sure. (I found another UX issue in Repoman enabled by the per-line comments, but I'll open a separate issue about it.)

isantop commented:
Yep, that's expected. Repolib has done special casing for paired deb/deb-src repos that have identical configuration otherwise. It should automatically dedup those internally. I'll get this merged so people's repoman is fixed
